File: leetcode/leetcode-java-solutions/AddBinary.java
class Solution {
    public String addBinary(String a, String b) {
        ArrayList<String> answerArr = new ArrayList<>();
        int carry = 0;
        if (a.length() > b.length()) {
            String longestString = a.length > b.length ? a : b;
            if (longestString == a) {
                b = "0".repeat(a.length()-b.length()) + b;
            } else {
                a = "0".repeat(b.length()-a.length()) + a;
            }
        }
        System.out.println(a);
        System.out.println(b);
        for (int i = a.length()-1; i>=0; i--) {
            switch(Integer.parseInt(a[i]) + Integer.parseInt(b[i]) {
                case 2:
                    if (carry == 1) {
                        answerArr.add(0, "1");
                        carry = 1
                    } else {
                        carry = 1;
                        answerArr.add(0, "0")
                    }
                    break
                case 0:
                    if (carry == 1) {
                        answerArr.add(0, "1");
                        carry = 0;
                    } else {
                        answerArr.add(0, "0");
                    }
                    break;
                case 1:
                    if (carry == 1) {
                        answerArr.add(0, "0");
                    } else {
                        answerArr.add(0, "1");
                    }
                    break;

            })
            if (i == 0 && carry == 1) {
                answerArr.add(0, "1");
            }
        }
        System.out.println(answerArr.join(""));
        return answerArr.join("");
    }
}

File: leetcode/leetcode-python-solutions/add-binary.py
def add_binary(a, b):
    answerArr = []
    carry = 0
    if len(a) != len(b):
        longest_string = a if len(a) > len(b) else b
        if longest_string == a:
            b = "0"*(len(a)-len(b)) + b
        else:
            a = "0"*(len(b)-len(a)) + a
    print(a)
    print(b)
    for i in range(len(a)-1, -1, -1):
        if int(a[i]) + int(b[i]) == 2:
            if carry == 1:
                answerArr.insert(0, "1")
            else:
                carry = 1
                answerArr.insert(0, "0")
        elif int(a[i]) + int(b[i]) == 0:
            if carry == 1:
                answerArr.insert(0, "1")
                carry = 0
            else:
                answerArr.insert(0, "0")
        elif int(a[i]) + int(b[i]) == 1:
            if carry == 1:
                answerArr.insert(0, "0")
            else:
                answerArr.insert(0, "1")
        if i == 0 and carry == 1:
            answerArr.insert(0, "1")

    print(answerArr)
    print("".join(answerArr))
    return "".join(answerArr)


add_binary("1", "111")


